### DISCUSSION TITLE
libgit2 v1.6.0

### DISCUSSION THREAD

**Discussion by ethomson** [libgit2/libgit2#6504] [Announcements]
v1.6
----

This is release v1.6.0, "Hubbeliges Krokodil". This release adds experimental SHA256 support and includes many new features and bugfixes.

### New features

* **Support for bare repositories with SHA256 support (experimental)** by @ethomson in https://github.com/libgit2/libgit2/pull/6191
   You can configure experimental SHA256 support in libgit2 with `cmake -DEXPERIMENTAL_SHA256=ON` during project setup. This is useful for considering future integrations, work on clients, and work on language bindings. At present, working with bare repositories should largely work, including remote operations. But many pieces of functionality - including working with the index - are not yet supported. As a result, **libgit2 with SHA256 support should not be used in production or released with package distribution.**

* **Support the notion of a home directory separately from global configuration directory** by @ethomson in https://github.com/libgit2/libgit2/pull/6455 and https://github.com/libgit2/libgit2/pull/6456
  Callers and language bindings can now configure the home directory that libgit2 uses for file lookups (eg, the `.ssh` directory). This configuration is separate from the git global configuration path.

* **stash: partial stash specific files** by @gitkraken-jacobw in https://github.com/libgit2/libgit2/pull/6330
  A stash can be created with only specific files, using a pathspec. This is similar to the `git stash push` command.

* **push: revparse refspec source, so you can push things that are not refs** by @sven-of-cord in https://github.com/libgit2/libgit2/pull/6362
  Pushes can be performed using refspecs instead of only references.

* **Support OpenSSL3** by @ethomson in https://github.com/libgit2/libgit2/pull/6464 and https://github.com/libgit2/libgit2/pull/6471
  OpenSSL 3 is now supported, both when compiled directly and dynamically loaded.

### Bug fixes
* winhttp: support long custom headers by @kcsaul in https://github.com/libgit2/libgit2/pull/6363
* Fix memory leak by @csware in https://github.com/libgit2/libgit2/pull/6382
* Don't fail the whole clone if you can't find a default branch by @torvalds in https://github.com/libgit2/libgit2/pull/6369
* #6366: When a worktree is missing, return `GIT_ENOTFOUND`. by @arroz in https://github.com/libgit2/libgit2/pull/6395
* commit-graph: only verify csum on `git_commit_graph_open()`. by @derrickstolee in https://github.com/libgit2/libgit2/pull/6420
* Ignore missing 'safe.directory' config during ownership checks by @kcsaul in https://github.com/libgit2/libgit2/pull/6408
* Fix leak in `git_tag_create_from_buffer` by @julianmesa-gitkraken in https://github.com/libgit2/libgit2/pull/6421
* http: Update httpclient options when reusing an existing connection. by @slackner in https://github.com/libgit2/libgit2/pull/6416
* Add support for `safe.directory *` by @csware in https://github.com/libgit2/libgit2/pull/6429
* URL parsing for google-compatible URLs by @ethomson in https://github.com/libgit2/libgit2/pull/6326
* Fixes #6433: `git_submodule_update` fails to update configured but missing submodule by @tagesuhu in https://github.com/libgit2/libgit2/pull/6434
* transport: fix capabilities calculation by @russell in https://github.com/libgit2/libgit2/pull/6435
* push: use resolved oid as the source by @ethomson in https://github.com/libgit2/libgit2/pull/6452
* Use `git_clone__submodule` to avoid file checks in workdir by @abizjak in https://github.com/libgit2/libgit2/pull/6444
* #6422: handle dangling symbolic refs gracefully by @arroz in https://github.com/libgit2/libgit2/pull/6423
* `diff_file`: Fix crash when freeing a patch representing an empty untracked file by @jorio in https://github.com/libgit2/libgit2/pull/6475
* clone: clean up options on failure by @ethomson in https://github.com/libgit2/libgit2/pull/6479
* stash: update strarray usage by @ethomson in https://github.com/libgit2/libgit2/pull/6487
* #6491: Sets `oid_type` on repos open with `git_repository_open_bare` by @arroz in https://github.com/libgit2/libgit2/pull/6492
* Handle Win32 shares by @ethomson in https://github.com/libgit2/libgit2/pull/6493
* Make failure to connect to ssh-agent non-fatal by @fxcoudert in https://github.com/libgit2/libgit2/pull/6497
* odb: don't unconditionally add `oid_type` to stream by @ethomson in https://github.com/libgit2/libgit2/pull/6499
* Pass hostkey & port to host verify callback by @fxcoudert in https://github.com/libgit2/libgit2/pull/6503

### Code cleanups
* meta: update version number to v1.6.0-alpha by @ethomson in https://github.com/libgit2/libgit2/pull/6352
* sha256: indirection for experimental functions by @ethomson in https://github.com/libgit2/libgit2/pull/6354
* Delete `create.c.bak` by @lrm29 in https://github.com/libgit2/libgit2/pull/6398
* Support non-cmake builds with an in-tree `experimental.h` by @ethomson in https://github.com/libgit2/libgit2/pull/6405

### Build and CI improvements
* tests: skip flaky-ass googlesource tests by @ethomson in https://github.com/libgit2/libgit2/pull/6353
* clar: remove ftrunacte from libgit2 tests by @boretrk in https://github.com/libgit2/libgit2/pull/6357
* CI Improvements by @ethomson in https://github.com/libgit2/libgit2/pull/6403
* fix compile on Windows with `-DWIN32_LEAN_AND_MEAN` by @christoph-cullmann in https://github.com/libgit2/libgit2/pull/6373
* Fixes #6365 : Uppercase windows.h include fails build in case-sensitive OS by @Vinz2008 in https://github.com/libgit2/libgit2/pull/6377
* ci: update version numbers of actions by @ethomson in https://github.com/libgit2/libgit2/pull/6448
* thread: avoid warnings when building without threads by @ethomson in https://github.com/libgit2/libgit2/pull/6432
* src: hide unused hmac() prototype by @0-wiz-0 in https://github.com/libgit2/libgit2/pull/6458
* tests: update clar test runner by @ethomson in https://github.com/libgit2/libgit2/pull/6459
* ci: always create test summaries, even on failure by @ethomson in https://github.com/libgit2/libgit2/pull/6460
* Fix build failure with `-DEMBED_SSH_PATH` by @vicr123 in https://github.com/libgit2/libgit2/pull/6374
* Define correct `off64_t` for AIX by @bzEq in https://github.com/libgit2/libgit2/pull/6376
* Fix some warnings in main by @ethomson in https://github.com/libgit2/libgit2/pull/6480
* strarray: remove deprecated declaration by @ethomson in https://github.com/libgit2/libgit2/pull/6486
* tests: always unset `HTTP_PROXY` before starting tests by @ethomson in https://github.com/libgit2/libgit2/pull/6498

### Documentation improvements
* add 2-clause BSD license to COPYING by @martinvonz in https://github.com/libgit2/libgit2/pull/6413
* Add new PHP bindings project to language bindings section of README.md by @RogerGee in https://github.com/libgit2/libgit2/pull/6473
* README: clarify the linking exception by @ethomson in https://github.com/libgit2/libgit2/pull/6494
* Correct the definition of "empty" in the docs for `git_repository_is_empty` by @timrogers in https://github.com/libgit2/libgit2/pull/6500

## New Contributors
* @christoph-cullmann made their first contribution in https://github.com/libgit2/libgit2/pull/6373
* @Vinz2008 made their first contribution in https://github.com/libgit2/libgit2/pull/6377
* @torvalds made their first contribution in https://github.com/libgit2/libgit2/pull/6369
* @derrickstolee made their first contribution in https://github.com/libgit2/libgit2/pull/6420
* @julianmesa-gitkraken made their first contribution in https://github.com/libgit2/libgit2/pull/6421
* @slackner made their first contribution in https://github.com/libgit2/libgit2/pull/6416
* @martinvonz made their first contribution in https://github.com/libgit2/libgit2/pull/6413
* @tagesuhu made their first contribution in https://github.com/libgit2/libgit2/pull/6434
* @russell made their first contribution in https://github.com/libgit2/libgit2/pull/6435
* @sven-of-cord made their first contribution in https://github.com/libgit2/libgit2/pull/6362
* @0-wiz-0 made their first contribution in https://github.com/libgit2/libgit2/pull/6458
* @abizjak made their first contribution in https://github.com/libgit2/libgit2/pull/6444
* @vicr123 made their first contribution in https://github.com/libgit2/libgit2/pull/6374
* @bzEq made their first contribution in https://github.com/libgit2/libgit2/pull/6376
* @gitkraken-jacobw made their first contribution in https://github.com/libgit2/libgit2/pull/6330
* @fxcoudert made their first contribution in https://github.com/libgit2/libgit2/pull/6497

<hr /><em>This discussion was created from the release <a href='https://github.com/libgit2/libgit2/releases/tag/v1.6.0'>libgit2 v1.6.0</a>.</em>

**Comment by ethomson** [comment]
Note that libgit2 v1.6.0 has been unpublished. I missed a version number update. As a result, [v1.6.1 has been published in its place](https://github.com/libgit2/libgit2/releases/tag/v1.6.1).

**Reply by giordano** [reply]
Side note, are there particularly reasons why the soversion was changed?  Did the API of any function changed in a breaking way?  Skimming through the release notes I don't see any breaking change, and tests in https://github.com/JuliaLang/julia/pull/48789 without changing anything in how we call libgit2.  Embedding the minor version number in the soversion means that you're breaking the ABI, which isn't fun for downstream packages if this is done unnecessarily.

**Comment by Batchyx** [comment]
PSA to packagers: in-source builds are broken in this release (it introduces conflicts with stdint.h), make sure your cmake integration uses a separate build directory.

**Reply by ethomson** [reply]
We don't recommend you `cmake` directly in the source root _but_ I also want it to work. I'm seeing a bunch of errors about `stdint.h` not being included when I:

```
cmake .
make
```

directly in the root of the libgit2 source tree. Just making sure that this is what you're describing?

**Reply by Batchyx** [reply]
Yes, this is what i am describing. This is due to 6c57bac6b18f6a30e1a8c8272a8dc367f202aa88 adding `${PROJECT_BINARY_DIR}/include/git2` in `src/util/CMakeLists.txt`. With in-source builds, this adds include/git2 as an include directory, so when the libc's `inttypes.h` tries to include `stdint.h`, `include/git2/stdint.h` is included instead and does not do anything on non-win32 (i'm not sure why 2bd3c80e7c303 exists, without it the problem would be clearer).

**Reply by ethomson** [reply]
This should be fixed with https://github.com/libgit2/libgit2/pull/6529
